# Time value of money (TVM)

Future value (FV) i.e. compounding

You save &#36;100 for 1 year at an interest rate of 3%
You save &#36;100 for 3 years at an interest rate of 3% p.a. (per annum)

$ FV = PV(1 + r)^n $

where
FV: Future Value
PV: Present Value (today)
r: Interest Rate (per period)
n: Number of periods

In [1]:
# 1 year
100 * (1 + 0.03) ** 1

103.0

In [2]:
# 3 years
100 * (1 + 0.03) ** 3

109.2727

Present value (PV) i.e. discounting

How many &#36; to save `today` at an interest rate of 4.5% p.a. to get &#36;110 in 1 year?
How many &#36; to save `today` at an interest rate of 4.5% p.a. to get &#36;110 in 3 years?

$ PV = \frac{FV}{(1 + r) ^ n} $

i.e. we have to discount the future value.

In [3]:
# 1 year discounted
110 / (1 + 0.045) ** 1

105.26315789473685

In [4]:
# 3 year discounted
110 / (1 + 0.045) ** 3

96.39262644604003

# Interest rates and returns

Today you receive the offer to deposit &#36;90 in a savings account, getting back &#36;93.50 in 1 year.
Today you receive the offer to deposit &#36;90 in a savings account, getting back &#36;93.50 in 3 years.

What is the interest rate?

$ r = (\frac{FV}{PV}) ^ \frac{1}{n} - 1 $

In [5]:
# 1 year
(93.50 / 90) ** (1 / 1) - 1

0.03888888888888897

In [6]:
# 3 years
(93.50 / 90) ** (1 / 3) - 1

0.012798463496144663

## Stock returns (price return)

1 year ago you invested &#36;50 in a stock that is now worth &#36;56.50.

$ r = \frac{P_{t + 1}}{P_t} - 1 $

where
$ {P_t} $: Price @ timestamp t
$ {P_{t + 1} $: Price @ t + 1
r: Period return (price return)

Return on your investment in a stock typically has 2 parts:
- any price increase
- any dividends

So now `r` would be:

$ r = \frac{P_{t + 1} + D_{t + 1}}{P_t} - 1 = \frac{P_{t + 1}}{P_t} - 1 + \frac{D_{t + 1}}{P_t} $

where
$ D_{t + 1} $: Dividend payment @ t + 1
r (is now): Period return (total return)

and on the right hand side of the formula:
$ \frac{P_{t + 1}}{P_t} - 1 $ is the (original) price return
and
$ \frac{D_{t + 1}}{P_t} $ is the dividend yield
added together gives the total return

In [7]:
# 1 year ago you invested $50 in a stock that is now worth $56.50
56.50 / 50 - 1

0.1299999999999999

In [8]:
# 1 year ago you invested $50 in a stock that paid a dividend of $2 and is now worth $56.50
(56.50 + 2.00) / 50 - 1

0.16999999999999993

# Using variables (where variables are "memory references")

In [9]:
# Future value: Invest $100 for 3 years at interest rate of 3%
fv = 100 * (1 + 0.03) ** 3

In [10]:
fv

109.2727

In [11]:
pv = 100
r = 0.03
n = 3

In [12]:
fv = pv * (1 + r) ** n

In [13]:
fv

109.2727

# print

In [14]:
print(56 / 50 - 1)

0.1200000000000001


In [15]:
print(56.50 / 50 - 1, (56.50 + 2) / 50 - 1)

0.1299999999999999 0.16999999999999993


In [16]:
print(56.50 / 50 - 1, (56.50 + 2) / 50 - 1, sep = ", ")

0.1299999999999999, 0.16999999999999993


In [17]:
print(56.50 / 50 - 1, (56.50 + 2) / 50 - 1, sep = "\n")

0.1299999999999999
0.16999999999999993


# Future Value (and Present Value) with many cashflows

Today you have &#36;100 in your savings account and you save another:
- &#36;10 in t1
- &#36;20 in t2
- &#36;50 in t3
- &#36;30 in t4
- &#36;25 in t5
(each cf (cashflow) at period's end)

Calculate the Future Value of your savings account after 5 years given an interest rate of 3% p.a.

$ FV_N = \sum_{t=0}^{N} CF_t * (1 + r)^{N - t} $

where
$ FV_N $: Future Value (at N)
$ CF_t $: Cashflow @ timestamp t
N: Total number of periods
r: Interest rate (per period)
t: timestamp (0, 1... N)

In [18]:
fv = 100 * (1 + 0.03) ** (5 - 0) + 10 * (1 + 0.03) ** (5 - 1) + 20 * (1 + 0.03) ** (5 - 2) + 50 * (1 + 0.03) ** (5 - 3) + 30 * (1 + 0.03) ** (5 - 4) + 25

In [19]:
fv

257.98203553

Let's improve the above awkward calculation.

In [20]:
cfs = [100, 10, 20, 50, 30, 25]

Normally we would index the above from left to right with indexes 0, 1 etc.
However, we can index from right to left starting with -1 with the following:
[100, 10, 20, 50, 30, 25]
 -6   -5  -4  -3  -2  -1
 e.g.

In [21]:
cfs[-2] # Will be 30

30

Let's recalculate the FV of our (multiple) cashflows by using a second list with the timestamps of the cashflows (a simple list of 0 to 5)

In [22]:
n = [0, 1, 2, 3, 4, 5]

f = 1 + 0.03

fv = cfs[0] * f ** n[-1] + cfs[1] * f ** n[-2] + cfs[2] * f ** n[-3] + cfs[3] * f ** n[-4] + cfs[4] * f ** n[-5] + 25

In [23]:
fv

257.98203553

And improve again:

In [24]:
sum([(cf * f ** i) for (cf, i) in list(zip(cfs, reversed(n)))])

257.98203553

In [25]:
# Or
sum([(cf * f ** i) for (i, cf) in list(enumerate(reversed(cfs)))])

257.98203552999996

And we'll do the PV equivalent of the above.

Today you agree on a payment plan that guarantees payouts of:
- &#36;50 in t1
- &#36;60 in t2
- &#36;70 in t3
- &#36;80 in t4
- &#36;100 in t5 (each cf at period's end)

Calculate the Funding amount / PV that needs to be paid into the plan today, t0 - Assume interest rate of 4% p.a.

In [26]:
cfs = [50, 60, 70, 80, 100]

f = 1 + 0.04

sum([(cf / f ** (1 + i)) for (i, cf) in list(enumerate(cfs))])

316.3570869232654

So, upon an initial investment of &#36;316.36 we receive the above cash flows from the bank.
i.e. if we have to pay &#36;316.36 at t0, this outgoing (for us) is a negative value, or
$ I_0 $ = -&#36;316.36

Formulating this:
$ PV = -I_0 $
or
$ I_0 + PV = 0 = NPV $

A real world example:
XYZ company evaluates to buy an additional machine that will increase future profits/cashflows by:
- &#36;20 in t1
- &#36;50 in t2
- &#36;70 in t3
- &#36;100 in t4
- &#36;50 in t5 (each cf at period's end)

The machine costs &#36;200 (investment in to).
Calculate the project's NPV and evaluate whether XYZ should persue the project.
XYZ's required rate of return (Cost of Capital) is 6% p.a.

$ NPV = I_0 + \sum_{t=1}^{N} \frac{CF_t}{(1 + r)^t} $

where
NPV: Net Present Value
$ I_0 $: Initial investment (negative)
$ CF_t $: Cashflow @ timestamp t
N: Total number of periods
r: Required interest rate return
t: timestamp (0, 1... N)

Unlike the previous example, $ I_0 \neq PV $ for a real world example.

> Simple decision rule:
>
> Accept the project if NPV > 0
>
> Reject the project if NPV < 0

> Interpretation of NPV:
>
> Total company value is the sum of all projects' NPVs

In [27]:
cfs = [-200, 20, 50, 70, 100, 50]

f = 1 + 0.06

sum([(cf / f ** i) for (i, cf) in list(enumerate(cfs))])

38.71337130837991

As the calculated NPV is positive we conclude that this is a good deal.
But would it still be a good deal if the purchase price was &#36;250?

In [28]:
sum([(cf / f ** i) for (i, cf) in list(enumerate([-250] + cfs[1:]))])

-11.286628691620088

# Data Types

In [29]:
type(100)

int

In [30]:
type(2.3)

float

In [31]:
cfs = [100, 50, 10]
type(cfs)

list

In [32]:
type(print)

builtin_function_or_method

In [33]:
fn = print

fn(cfs)

[100, 50, 10]


# List slicing

In [34]:
cfs = [-200, 20, 50, 70, 100, 50]

In [35]:
cfs[0 : 1]

[-200]

In [36]:
cfs[1 : 4]

[20, 50, 70]

In [37]:
cfs[: 2]

[-200, 20]

In [38]:
cfs[:]

[-200, 20, 50, 70, 100, 50]

In [39]:
cfs[: -1]

[-200, 20, 50, 70, 100]

In [40]:
cfs[-3 :]

[70, 100, 50]

In [43]:
cfs[: : 2] # All elements but only every 2nd element, this is equivalent to cfs[0 : 6 : 2]

[-200, 50, 100]

In [42]:
cfs[1 : 5 : 2] # 2 means take steps of size 2

[20, 70]

In [44]:
cfs[: : -1] # This is the equivalent of reversing a list

[50, 100, 70, 50, 20, -200]